File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.java
import java.util.*;
import java.io.*;

public class Main {
    static int[][] gp;
    static int[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] line = br.readLine().split(" ");
        int N = Integer.parseInt(line[0]);              // node 개수
        int M = Integer.parseInt(line[1]);              // 간선 개수

        gp = new int[N + 1][N + 1];
        visited = new int[N + 1];
        for (int m = 0; m < M; m++) {
            line = br.readLine().split(" ");
            int x = Integer.parseInt(line[0]);
            int y = Integer.parseInt(line[1]);
            gp[x][y] = 1;
            gp[y][x] = 1;
        }

        int cnt = 0;
        for (int n = 1; n <= N; n++) {
            if (visited[n] == 0) {
                bfs(n);
                cnt ++;
            }
        }
        System.out.println(cnt);
    }

    static void bfs(int N) {
        visited[N] = 1;
        for (int i = 1; i < gp.length; i++) {
            if (gp[N][i] == 1 && visited[i] == 0) bfs(i);
        }
    }
}

File: 백준/Silver/11724. 연결 요소의 개수/연결 요소의 개수.py
import sys
from collections import deque
sys.setrecursionlimit(10 ** 5)

input = __import__('sys').stdin.readline

def dfs(node):
    for nxt in range(1, N + 1):
        if gp[node][nxt] == 1 and visited[nxt] != 1:
            visited[nxt] = 1
            dfs(nxt)
    return

def bfs(node):
    que = deque()
    que.append(node)
    visited[node] = 1
    while len(que) > 0:
        now = que.popleft()
        for next in range(1, N + 1):
            if gp[now][next] == 1 and visited[next] != 1:
                visited[next] = 1
                que.append(next)
    return

N, M= map(int, input().split())
gp = [[0 for _ in range(N + 1)] for _ in range(N + 1)]
for _ in range(M):
    x, y = map(int,input().split())
    gp[x][y] = 1
    gp[y][x] = 1

visited = [0] * (N+1)
cnt = 0
for n in range(1, N + 1):
    if visited[n] == 0:
        visited[n] = 1
        # dfs(n)
        bfs(n)
        cnt += 1
print(cnt)
